common/timer.go

```diff
@@ -152,74 +185,114 @@ func (tstore *TimerStore) setNextTimeout(ctx sdk.Context, which types.TimerType,
 	store := tstore.getStoreTimer(ctx, which)
-	store.Set(types.EncodeKey(value), []byte(data))
+	timerKey := types.EncodeBlockAndKey(value, key)
+	store.Set(timerKey, data)
 
 	nextValue := tstore.getNextTimeout(ctx, which)
 	if value < nextValue {
 		tstore.setNextTimeout(ctx, which, value)
 	}
 }
 
-func (tstore *TimerStore) delTimer(ctx sdk.Context, which types.TimerType, value uint64) {
+func (tstore *TimerStore) delTimer(ctx sdk.Context, which types.TimerType, value uint64, key []byte) {
 	store := tstore.getStoreTimer(ctx, which)
-	store.Delete(types.EncodeKey(value))
+	timerKey := types.EncodeBlockAndKey(value, key)
+	if !store.Has(timerKey) {
+		panic(fmt.Sprintf("delTimer which %d block %d key %v: no such timer", which, value, key))
+	}
+	store.Delete(timerKey)
 }
 
-func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, data string) {
-	tstore.addTimer(ctx, types.BlockHeight, block, data)
+// AddTimerByBlockHeight adds a new timer to expire on a given block height.
+// If a timer for that <block, key> tuple exists, it will be overridden.
+func (tstore *TimerStore) AddTimerByBlockHeight(ctx sdk.Context, block uint64, key []byte, data []byte) {
+	if block <= uint64(ctx.BlockHeight()) {
+		panic(fmt.Sprintf("timer expiry block %d smaller than ctx block %d",
+			block, uint64(ctx.BlockHeight())))
+	}
+	tstore.addTimer(ctx, types.BlockHeight, block, key, data)
 }
 
-func (tstore *TimerStore) AddTimerByBlockTime(ctx sdk.Context, timestamp uint64, data string) {
-	tstore.addTimer(ctx, types.BlockTime, timestamp, data)
+// AddTimerByBlockTime adds a new timer to expire on a future block with the given timestamp.
+// If a timer for that <timestamp, key> tuple exists, it will be overridden.
+func (tstore *TimerStore) AddTimerByBlockTime(ctx sdk.Context, timestamp uint64, key []byte, data []byte) {
+	if timestamp <= uint64(ctx.BlockTime().UTC().Unix()) {
+		panic(fmt.Sprintf("timer expiry time %d smaller than ctx time %d",
+			timestamp, uint64(ctx.BlockTime().UTC().Unix())))
+	}
+	tstore.addTimer(ctx, types.BlockTime, timestamp, key, data)
 }
 
-type timerTuple struct {
-	value uint64
-	data  string
+// DelTimerByBlockHeight removes an eximsting timer for the <block, key> tuple.
+func (tstore *TimerStore) DelTimerByBlockHeight(ctx sdk.Context, block uint64, key []byte) {
+	tstore.delTimer(ctx, types.BlockHeight, block, key)
 }
 
-func (tstore *TimerStore) tickValue(ctx sdk.Context, which types.TimerType, tickValue uint64) {
-	nextValue := tstore.getNextTimeout(ctx, which)
-	if tickValue < nextValue {
-		return
+// DelTimerByBlockHeight removes an eximsting timer for the <timetamp, key> tuple.
+func (tstore *TimerStore) DelTimerByBlockTime(ctx sdk.Context, timestamp uint64, key []byte) {
+	tstore.delTimer(ctx, types.BlockTime, timestamp, key)
+}
+
+func (tstore *TimerStore) DumpTimers(ctx sdk.Context, which types.TimerType) string {
```

Comment:
DumpTimers is a pretty good name, but maybe change it to PrintTimers (which is more clear)?

Reply:
it's a common convention (like core-dump). 
but can rename, sure.